Conditional form field

kuzmany opened the pull request:
**Please be sure you are submitting this against the _staging_ branch.**

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | 
| New feature? | 
| Automated tests included? |
| Related user documentation PR URL | 
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | Closes https://github.com/mautic/mautic/issues/8595
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR added conditional fields to form.
Every selectbox now attach fields displayed related to selected values. Support also multiselect values.

![image](https://user-images.githubusercontent.com/462477/81265914-2881ea00-9044-11ea-9e3b-e5dc597cdada.png)

![image](https://user-images.githubusercontent.com/462477/81265947-36d00600-9044-11ea-89e3-f6f8e44e87d4.png)

![captured (5)](https://user-images.githubusercontent.com/462477/80629176-c0834080-8a52-11ea-9ffe-95cb00b67803.gif)


#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Don't forget refresh browser due CSS changes before test.
3. Create form
4. Create Select field type
5. After save you are able to add field displayed only if visitor select values
![image](https://user-images.githubusercontent.com/462477/80627977-fde6ce80-8a50-11ea-9672-626ea5136ec2.png)
6. Don't forget set condition tab for each child field
![image](https://user-images.githubusercontent.com/462477/80628059-19ea7000-8a51-11ea-9ef1-e80c2e3375d5.png)
7. Then go to preview and test form
8. Test also embeded JS and Manual copy version of form
9. Also test required and contact matching for child fields - take care If hidden fields wasn't store or set to contact profila.
10. Also test send form results action, pagination etc
11. Also test some of progressive profiling feature

kuzmany committed:
Init commit

kuzmany committed:
Add conditions column to form_fields

kuzmany committed:
Adda ajax lookup for conditional options

kuzmany committed:
Adda ajax choosen for conditional field

kuzmany committed:
Added info about conditions to panel

kuzmany committed:
Start working on parent/child structure of conditional fields

kuzmany committed:
Another work for conditional fields

kuzmany committed:
Another part work on conditional field backend

kuzmany committed:
Rework backend conditinal fields with parent as string

kuzmany committed:
Added showOn

kuzmany committed:
Redesign backend for conditional fields

kuzmany committed:
Minor

kuzmany committed:
Add showForConditionalField to skip field parent not want it

kuzmany committed:
Add support for multi select values

kuzmany committed:
Add field matching and validation for conditional fields

kuzmany committed:
Fix ordering for nonparent/parent

kuzmany committed:
Add PropertiesAccessor to access to properties

kuzmany committed:
Delete children fields after parent remove

kuzmany committed:
Add field save/delete events for developers

kuzmany committed:
Rework fields styling

kuzmany committed:
Minor fixes

kuzmany committed:
Just one level conditioanl at the moment

kuzmany committed:
Add support for Any selected value

npracht commented:
- change: "Show only if choose from (field) values (option)" to "Show only if selection from (field) equals/includes/excludes (option/s)"
- the option to add a subfield is not well design (too big and by default shown), could we change to something smoother like that ?
![Capture d’écran 2020-04-30 à 17 01 12](https://user-images.githubusercontent.com/14075239/80726082-3dbabe00-8b04-11ea-8159-d55b9acf7c5c.png)
- we need a deeper condition selector:
![Capture d’écran 2020-04-30 à 17 08 05](https://user-images.githubusercontent.com/14075239/80726804-2e884000-8b05-11ea-9676-16ec72ecc5a1.png)

kuzmany commented:
what is different between equal and include if you can choose multiple values?
What is benefit for exclude If you can select multiple values to show (make same  results)?

npracht commented:
https://mautibox.com/8660/_profiler/647407?panel=logger
![image](https://user-images.githubusercontent.com/14075239/80835003-32de5700-8bf2-11ea-80e4-a3c94eb09cbe.jpeg)

kuzmany committed:
Fix syncList from contact field

kuzmany committed:
Redesign link to create new sub field

kuzmany committed:
Minor CSS changes

kuzmany committed:
Fix reordering and improve form details  field list

kuzmany committed:
Minor to style/css

kuzmany commented:
@npracht bug fixed
I don't see benefits from include/exclude, because with current feature you can achieve all of that. Other options make feature more complex and probability of bugs
I've also add option: Any selected value (for example Color of car would displayed on select any car brand)

I've improve few another things, I hope we can move to Ready to test state

npracht commented:
![Capture d’écran 2020-05-04 à 14 26 53](https://user-images.githubusercontent.com/14075239/80965709-52a09580-8e13-11ea-90b5-8a6f2fc00d4e.png)
When I click on this link, it refresh the `/edit` page instead of opening the conditional field.

kuzmany commented:
@npracht try refresh cache.
Should work like this (tested on Mautibox)

![captured (6)](https://user-images.githubusercontent.com/462477/80966124-2b969380-8e14-11ea-8d75-f85007fc1135.gif)

kuzmany committed:
Delete form-field.html.php

atowne-mautic commented:
@kuzmany I've been working with Norman a bit on this in the background, but happy to actually join in with the community here :)

To answer your earlier question about including/excluding, the excluding option is very helpful if there's a long list of options for the 'parent' field, and the user only wants to show the conditional field for **most** of the options. It's not required to have excluding as an option, but would improve the user experience. I don't think there's a significant difference between equals and 'including' with only 1 option selected.

In Mautibox, I tested adding a condition to the List - Country field and got the same error @npracht reported (refreshes the edit page). But I was able to add a checkbox and a radio group for different options on a select field (see form ID 6). 

I also see that the change mentioned above doesn't seem to be here: change "Show only if choose from (field) values (option)" to "Show only if selection from (field) equals/includes/excludes (option/s)". Not sure if that was just a suggestion that hasn't been done yet, or may have been missed with another update.

One other request/suggestion, is to be able to add multiple levels. For example (using my fields in form 6), if someone chooses Burgers on the American Food question (which is dependent on the first question being answered with "American", I'd like to be able to ask another question for if someone likes cheese on their burgers. This had been written into the code I'd shared with Norman.

Finally, just checking - is this being tested against M3? I see Mautibox is on 2.16.2.

npracht commented:
@atowne-mautic concerning Mautic version, we do it in 2 steps.

kuzmany committed:
Add toggle to show/hidden sub field select

kuzmany committed:
Add include/exclude

kuzmany committed:
Add support for checkbox and radio groups

kuzmany commented:
@atowne-mautic I've added support for include/exclude expression. Also fixed support  conditional fields for checkbox/radio groups
Now your test should work properly https://mautibox.com/8660/s/forms/preview/6

atowne-mautic commented:
@kuzmany Thanks for that! The form preview looks like it works, but now I see this on the edit page: 
![image](https://user-images.githubusercontent.com/46789594/81184935-fa86a200-8f7e-11ea-9c54-72c72c2f02f0.png)

kuzmany committed:
Fix radio/checkbox groups in backend

kuzmany commented:
@atowne-mautic fixed with last commit

alanhartless commented:
@kuzmany @npracht would you guys be willing for us to tag team on making this M3 compatible?

atowne-mautic commented:
@kuzmany back to normal now, thanks. Would the same code you used to fix the issue causing the edit page to refresh also fix the same issue for the country field?

One additional (and critical) piece Korcomptenz had previously written was to enable mapping to the same contact field multiple times on the same form **if** that contact field were dependent on another field. The key example is country/state - if a contact is in the US, I want to show US states and map to the State contact field. If the contact is in Canada, I want to show Canadian provinces and map to the State field, etc. Is that already supported here? I can't quite test that yet.

And another teeny little UI thing, we're just missing the left border on the dependency/mapping bar below a field. Not critical, but would make this look a little more polished:
![image](https://user-images.githubusercontent.com/46789594/81185972-4423bc80-8f80-11ea-9a0f-a14076699085.png)
(the right border is there, I just didn't take a full-width screenshot)

kuzmany commented:
@atowne-mautic country/state relation need different logic and not ready yet.

kuzmany committed:
Fix country selector

kuzmany committed:
Replace toggle to show for child  selector

escopecz committed:
Migration that adds form_fields.parent_id field

escopecz committed:
Fixing PHP Notice - Undefined index: empty_value

escopecz committed:
Shorter syntax used

escopecz commented on app/bundles/FormBundle/Event/FormFieldEvent.php:
```diff
@@ -0,0 +1,48 @@
+use Mautic\CoreBundle\Event\CommonEvent;
+use Mautic\FormBundle\Entity\Field;
+
+class FormFieldEvent extends CommonEvent
```
@kuzmany is there any plan with this event? I don't see it used in any subscriber.

kuzmany replied:
Yes, I wanted to use and play with delete children fields. But then not used it. But I decide let it there, because was already written and maybe some  devs could use it in future

kuzmany replied:
I've also try use foreign keys for parent_id to id but there was issue with cascade processing during the processing form (add/edit/delete). I was not able to figure out.

escopecz replied:
That's what I thought but wanted to make sure I haven't overlooked anything. I'll update it a little bit with some defensive programming techniques since it's a new class so we have an opportunity to create it the right way and don't worry about BC breaks.

kuzmany replied:
Don't worry. We have branch for M2 and all your changes I will cherry pick there except M3 stuff. Thanks

escopecz replied:
No need. The changes I did so far are compatible with M2 and M3. No need for cherry-picking. Once we are ready with bug fixes and complete the features then we can take this branch, clone it and make the final M3 touches there.

kuzmany replied:
👍

escopecz committed:
Preparing the subscriber for M3 compatibility

Also removing unnecessary comments, applyging defensive programming technique

escopecz committed:
Removing unecessary abstract class and using the interface directly

Also adding final as it's a value object and doesn't need to be mocked in tests. Adding missing getter so all the properties from the constructor could be used, removing unecessary comments

escopecz committed:
Define class properties

escopecz commented:
I'm stuck whole day on this:
1. Create a new form
2. Create a country field
3. Add a condition select field linked with contact state field to the country field. Select Yes that you want to use options from the contact field. No other configuration.
4. Save the form
5. Edit the condition field. It runs out of memory:
```
[2020-05-11 15:54:32] mautic.CRITICAL: Uncaught PHP Exception Symfony\Component\Debug\Exception\OutOfMemoryException: "Error: Allowed memory size of 1073741824 bytes exhausted (tried to allocate 66195456 bytes)" at /var/www/html/app/bundles/CoreBundle/Views/FormTheme/Custom/sortablelist_row.html.php line 29 {"exception":"[object] (Symfony\\Component\\Debug\\Exception\\OutOfMemoryException(code: 0): Error: Allowed memory size of 1073741824 bytes exhausted (tried to allocate 66195456 bytes) at /var/www/html/app/bundles/CoreBundle/Views/FormTheme/Custom/sortablelist_row.html.php:29)
[stacktrace]
#0 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(175): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->()
#1 /var/www/html/vendor/symfony/templating/PhpEngine.php(78): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->evaluate()
#2 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(144): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render()
#3 /var/www/html/vendor/symfony/form/Extension/Templating/TemplatingRendererEngine.php(37): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render()
#4 /var/www/html/vendor/symfony/form/FormRenderer.php(268): Symfony\\Component\\Form\\Extension\\Templating\\TemplatingRendererEngine->renderBlock()
#5 /var/www/html/vendor/symfony/framework-bundle/Templating/Helper/FormHelper.php(171): Symfony\\Component\\Form\\FormRenderer->searchAndRenderBlock()
#6 /var/www/html/app/bundles/FormBundle/Views/Builder/field.html.php(231): Mautic\\CoreBundle\\Templating\\Helper\\FormHelper->row()
#7 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(175): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->()
#8 /var/www/html/vendor/symfony/templating/PhpEngine.php(78): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->evaluate()
#9 /var/www/html/app/bundles/CoreBundle/Templating/Engine/PhpEngine.php(144): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render()
#10 /var/www/html/vendor/symfony/templating/DelegatingEngine.php(41): Mautic\\CoreBundle\\Templating\\Engine\\PhpEngine->render()
#11 /var/www/html/vendor/symfony/framework-bundle/Controller/Controller.php(163): Symfony\\Bundle\\FrameworkBundle\\Templating\\DelegatingEngine->render()
#12 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(393): Mautic\\FormBundle\\Controller\\FieldController->renderView()
#13 /var/www/html/app/bundles/FormBundle/Controller/FieldController.php(329): Mautic\\FormBundle\\Controller\\FieldController->ajaxAction()
#14 /var/www/html/app/bundles/CoreBundle/Controller/CommonController.php(474): Mautic\\FormBundle\\Controller\\FieldController->editAction()
#15 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(135): Mautic\\FormBundle\\Controller\\FieldController->executeAction()
#16 /var/www/html/vendor/symfony/http-kernel/HttpKernel.php(57): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handleRaw()
#17 /var/www/html/vendor/symfony/http-kernel/DependencyInjection/ContainerAwareHttpKernel.php(67): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handle()
#18 /var/www/html/vendor/symfony/http-kernel/Kernel.php(183): Symfony\\Component\\HttpKernel\\DependencyInjection\\ContainerAwareHttpKernel->handle()
#19 /var/www/html/app/AppKernel.php(141): AppKernel->handle()
#20 /var/www/html/app/middlewares/CORSMiddleware.php(93): AppKernel->handle()
#21 /var/www/html/app/middlewares/CatchExceptionMiddleware.php(45): Mautic\\Middleware\\CORSMiddleware->handle()
#22 /var/www/html/app/middlewares/Dev/IpRestrictMiddleware.php(66): Mautic\\Middleware\\CatchExceptionMiddleware->handle()
#23 /var/www/html/app/middlewares/VersionCheckMiddleware.php(57): Mautic\\Middleware\\Dev\\IpRestrictMiddleware->handle()
#24 /var/www/html/vendor/stack/builder/src/Stack/StackedHttpKernel.php(23): Mautic\\Middleware\\VersionCheckMiddleware->handle()
#25 /var/www/html/vendor/stack/run/src/Stack/run.php(13): Stack\\StackedHttpKernel->handle()
#26 /var/www/html/index_dev.php(33): Stack\
un()
#27 /var/www/html/index_dev.php(0): {main}()
#28 {main}
"} 
```
@kuzmany do you have any clue why? It seems like it tries to add an array of countries to the label, but I can't find where it is happening.

alanhartless commented:
Not sure how long this will be available but I can reproduce in https://mautibox.com/8660/s/forms/edit/10 (try editing the state field)

alanhartless commented:
Oh, https://mautibox.com/8660/s/forms/edit/6 has a custom list of options. So this fails to work when syncing the contact field's state field with the form field on the properties tab. And seems this is happening on Mautic master so not caused by this PR.

escopecz commented:
I found out what's going on with fresh head. The problem happens only in the dev environment and it's triggered by this PHP notice:
```
Symfony\Component\Debug\Exception\ContextErrorException: PHP Notice - Array to string conversion - in file /var/www/html/app/bundles/CoreBundle/Views/FormTheme/Custom/form_widget_simple.html.php - at line 44
```
Every time this notice is logged in the dev log it will also print all the values from the state field. It generates 76MB of logs on one request and then it fails on memory limit which is 1GB in my case. So now I have to find out why we are using that template to render an array value.

kuzmany commented:
@escopecz That's strange. I did test with selectboxes (+ multiselectboxes) and manual values or sync values from contact field. I didn't notice that issue.

escopecz commented:
The problem with the state field is that it's a multidimensional array. It shows countries as opt groups and states as options. Sortable templates don't support that. As Alan mentioned, it's not an issue caused by this PR.

escopecz committed:
The data-mautic-form-show-on param wasn't set if the $fields weren't passed to the field_helper.

escopecz committed:
Oops, the radio field already was passing the fields

escopecz committed:
Cover all posibilities of Field::showForConditionalField() with tests

escopecz committed:
Refactoring to return early

escopecz committed:
CS fixes

escopecz committed:
FormFieldEventTest added

escopecz committed:
Namespace fix for a test

escopecz committed:
FormConditionalSubscriber tested

escopecz committed:
FormFieldConditionType tested

escopecz commented on app/bundles/FormBundle/Helper/PropertiesAccessor.php:
```diff
@@ -0,0 +1,106 @@
+                } elseif (isset($option['label']) && isset($option['value'])) {
+                    $choices[$option['value']] = $option['label'];
+                } else {
+                    foreach ($option as $opt) {
```
@kuzmany I'm writing a unit test for this else part, but I can't find the right array format that could get here. Can you please provide me some example?

escopecz committed:
Removing unused dependency

escopecz committed:
PropertiesAccessor test

escopecz committed:
Changes in FormModel tested

escopecz commented on app/bundles/FormBundle/Helper/PropertiesAccessor.php:
```diff
@@ -0,0 +1,91 @@
+        foreach ($options as $option) {
+            if (is_array($option)) {
+                if (isset($option['label']) && isset($option['alias'])) {
+                    $choices[$option['alias']] = $option['label'];
```
This part as well. If it doesn't have 'value' then it won't get to this foreach (line 49). For what array structure was this built?

kuzmany replied:
This should cover fieldsprocessing, when I wanted to build choices of all  selectboxes from form etc. 
But I think it was used in first version, and then I remove this feature.

kuzmany replied:
Maybe doesn't exists. In first version I've tried process any properties from form fields, or lead fields. I found a few formats (bools, | divided, arrays). Maybe it was country/state field from sync from Lead matching field.

escopecz committed:
Removing the failing test for now. Discussing with the author how to cover the code

escopecz commented:
The Pipedrive functional tests started to fail on the staging branch as well. I'll create a M3 clone of this PR to see if it will be a problem there too. Other than the 2 questions above I have no more todos for this PR.

escopecz commented:
M3 PR clone: https://github.com/mautic/mautic/pull/8794

kuzmany committed:
Implementing special behaviour for opt group conditional select field to make the country-state combo simple

kuzmany committed:
Stop display child If parent is hidden type

kuzmany committed:
Add fixes from M3

165c4b2, efbc167, 13601dc, 87c19af

kuzmany committed:
Add support for checkboxes

kuzmany committed:
Added support for radiogrp

kuzmany commented:
Close favor to https://github.com/mautic/mautic/pull/8794

kuzmany commented:
Already exists version for M3 https://github.com/mautic/mautic/pull/8794 
Maybe we should close
